### PR TITLE
Implement Logical/Execution plan for HashJoins.

### PR DESCRIPTION
**Pull request by matriv**

### CODE REVIEW

File: sql/src/main/java/io/crate/execution/engine/join/HashJoinOperation.java [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+                    nlResultConsumer.accept(joinIterator, null);
+                } else {
+                    nlResultConsumer.accept(null, failure);
+                }
+            });
+    }
+
+    private static Function<Row, Integer> getHashBuilderFromSymbols(InputFactory inputFactory, List<Symbol> inputs) {
+        InputFactory.Context<? extends CollectExpression<Row, ?>> leftCtx = inputFactory.ctxForInputColumns(inputs);
+        Object[] leftValues = new Object[leftCtx.topLevelInputs().size()];
```
[the remark this answers is not in the document]
Comment (matriv) [reply sql/src/main/java/io/crate/execution/engine/join/HashJoinOperation.java]: yep sorry, was extracted as method...

File: sql/src/main/java/io/crate/execution/dsl/phases/JoinPhase.java [hunk 2]
```diff
@@ -25,65 +25,81 @@
...
 
     private final Collection<String> executionNodes;
     private final MergePhase leftMergePhase;
     private final MergePhase rightMergePhase;
     private final JoinType joinType;
+    private final ExecutionPhase.Type type;
 
     @Nullable
     private final Symbol joinCondition;
+    private final List<Symbol> leftJoinConditionInputs;
```
Comment (mfussenegger) [inline sql/src/main/java/io/crate/execution/dsl/phases/JoinPhase.java]: I'm not sure if it's a good idea to add properties that are only relevant for certain types of this phase. Wouldn't it be cleaner to add a dedicated HashJoinPhase?

File: sql/src/main/java/io/crate/execution/dsl/phases/JoinPhase.java [hunk 3]
```diff
@@ -149,15 +177,50 @@ public NestedLoopPhase(StreamInput in) throws IOException {
         } else {
             joinCondition = null;
         }
+
+        int numLeftJoinInput = in.readVInt();
+        if (numLeftJoinInput > 0) {
```
Comment (mfussenegger) [inline sql/src/main/java/io/crate/execution/dsl/phases/JoinPhase.java]: Might want to make use of `Symbols.listFromStream` and `Symbols.toStream`

File: sql/src/main/java/io/crate/execution/dsl/phases/MergePhase.java [hunk 4]
```diff
@@ -59,20 +62,57 @@
...
+
+    public static boolean isNeeded(Collection<String> executionNodes,
+                                   ResultDescription resultDescription,
+                                   boolean isDistributed) {
+        return isDistributed ||
+               resultDescription.hasRemainingLimitOrOffset() ||
+               !resultDescription.nodeIds().equals(executionNodes);
+    }
+
+    public static MergePhase buildForJoin(PlannerContext plannerContext,
```
Comment (mfussenegger) [inline sql/src/main/java/io/crate/execution/dsl/phases/MergePhase.java]: Is this `Join` specific? If so why, and I think if it is we should avoid putting it into `MergePhase`, as the MergePhase should be kept unaware of joins.

File: sql/src/main/java/io/crate/execution/dsl/phases/MergePhase.java [hunk 5]
```diff
@@ -59,20 +62,57 @@
...
+                                          ResultDescription resultDescription,
+                                          Collection<String> eecutionNodes) {
```
Comment (mfussenegger) [inline sql/src/main/java/io/crate/execution/dsl/phases/MergePhase.java]: typo in executionNodes

File: sql/src/main/java/io/crate/execution/dsl/phases/MergePhase.java [hunk 6]
```diff
@@ -59,20 +62,57 @@
     /**
      * expects sorted input and produces sorted output
      */
-    @Nullable private PositionalOrderBy positionalOrderBy;
+    @Nullable
+    private PositionalOrderBy positionalOrderBy;
+
+    public static boolean isNeeded(Collection<String> executionNodes,
+                                   ResultDescription resultDescription,
+                                   boolean isDistributed) {
```
Comment (mfussenegger) [inline sql/src/main/java/io/crate/execution/dsl/phases/MergePhase.java]: same here, the `isDistributed` is currently rather specific to joins - I think we should try to avoid putting it here - especially without any sort of documentation that explains the flag

File: sql/src/main/java/io/crate/execution/dsl/phases/MergePhase.java [hunk 7]
```diff
@@ -59,20 +62,57 @@
...
-     *                  the number of upstream phases.
-     * @param executionNodes The nodes where this MergePhase executes.
-     * @param inputTypes The types of the input rows.
-     * @param projections The projections to apply when merging.
-     * @param distributionInfo The default strategy to use when distributing the results of the MergePhase.
+     *
+     * @param jobId             The JobID of the entire execution.
+     * @param executionNodeId   A unique execution id for this phase.
+     * @param name              The name of the MergePhase.
+     * @param numUpstreams      The number of nodes to expect data from.
```
Comment (mfussenegger) [inline sql/src/main/java/io/crate/execution/dsl/phases/MergePhase.java]: Unrelated formatting change?

File: sql/src/main/java/io/crate/execution/engine/join/JoinOperation.java [hunk 8]
```diff
@@ -0,0 +1,69 @@
...
+
+import io.crate.concurrent.CompletionListenable;
+import io.crate.data.BatchIterator;
+import io.crate.data.Row;
+import io.crate.data.RowConsumer;
+
+import javax.annotation.Nullable;
+import java.util.concurrent.CompletableFuture;
+
+public abstract class JoinOperation implements CompletionListenable {
```
Comment (mfussenegger) [inline sql/src/main/java/io/crate/execution/engine/join/JoinOperation.java]: I think it would be nice if we could either try to avoid the inheritance here; or make this abstract class more concrete in what it's responsible for. Right now it looks like a random collection of functionality that happens to be shared.

 `getBatchConsumer` could be static and then all that's left here is 3 attributes.

File: sql/src/test/java/io/crate/planner/operators/JoinTest.java [hunk 9]
```diff
@@ -89,10 +92,67 @@ public void testTablesAreSwitchedIfLeftIsSmallerThanRight() throws Exception {
...
+        PlannerContext context = e.getPlannerContext(clusterService.state());
+        LogicalPlanner logicalPlanner = new LogicalPlanner(functions, tableStats);
+        SubqueryPlanner subqueryPlanner = new SubqueryPlanner((s) -> logicalPlanner.planSubSelect(s, context));
+        SessionContext sessionContext = SessionContext.create();
+        sessionContext.setHashJoinEnabled(true);
+        LogicalPlan operator = JoinPlanBuilder.createNodes(mss, mss.where(), subqueryPlanner,  sessionContext)
+            .build(tableStats, Collections.emptySet());
+        assertThat(operator, instanceOf(HashJoin.class));
+        assertThat(((HashJoin) operator).leftRelation.toString(), is("QueriedTable{DocTableRelation{doc.users}}"));
+        assertThat(((HashJoin) operator).rightRelation.toString(), is("QueriedTable{DocTableRelation{doc.locations}}"));
```
[the remark this answers is not in the document]
Comment (matriv) [reply sql/src/test/java/io/crate/planner/operators/JoinTest.java]: hm, kind of disagree, I have those to still check that the relations are the same (left/right) before building the execution plan. What do you think?

File: sql/src/main/java/io/crate/planner/operators/Join.java [hunk 10]
```diff
@@ -0,0 +1,123 @@
...
+    public Map<LogicalPlan, SelectSymbol> dependencies() {
+        HashMap<LogicalPlan, SelectSymbol> deps = new HashMap<>();
+        deps.putAll(lhs.dependencies());
+        deps.putAll(rhs.dependencies());
+        return deps;
+    }
+
+    @Override
+    public LogicalPlan tryOptimize(@Nullable LogicalPlan pushDown) {
+        if (pushDown instanceof Order) {
```
Comment (mfussenegger) [inline sql/src/main/java/io/crate/planner/operators/Join.java]: Again a bit of nitpicking against inheritance. I think we could extract the tryOptimize logic into a static functions within a `JoinOperators` and then avoid this Join base class. The operator class hierarchy is already at 4 levels - I would appreciate it if we could avoid making it bigger.

The `JoinOperators` class could also contain the helper functions that are now in the `MergePhase` which don't really belong there.

File: sql/src/test/java/io/crate/planner/operators/JoinTest.java [hunk 11]
```diff
@@ -89,10 +92,67 @@ public void testTablesAreSwitchedIfLeftIsSmallerThanRight() throws Exception {
...
-        nl = (NestedLoop) operator.build(context, projectionBuilder, -1, 0, null, null, Row.EMPTY, emptyMap());
+        nl = (Join) operator.build(context, projectionBuilder, -1, 0, null, null, Row.EMPTY, emptyMap());
         assertThat(
             ((Collect) nl.left()).collectPhase().distributionInfo().distributionType(),
             is(DistributionType.SAME_NODE)
         );
     }
+
+    @Test
+    public void testHashJoin() throws Exception {
```
Comment (mfussenegger) [inline sql/src/test/java/io/crate/planner/operators/JoinTest.java]: maybe the method name can be a bit more concrete? what about HashJoin is tested?

File: sql/src/test/java/io/crate/planner/SelectPlannerTest.java [hunk 12]
```diff
@@ -704,4 +705,12 @@ public void testFilterOnPKSubsetResultsInPKLookupPlanIfTheOtherPKPartIsGenerated
             "Get[doc.t_pk_part_generated | 1 | DocKeys{0, 0}"
         ));
     }
+
+    @Test
+    public void testInnerEquiJoinWithHashJoin() throws Exception {
```
Comment (mfussenegger) [inline sql/src/test/java/io/crate/planner/SelectPlannerTest.java]: maybe rename to clarify the test scenario - e.g. `testInnerJoinResultsInHashJoinIfHashJoinIsEnabled`

File: sql/src/main/java/io/crate/execution/dsl/phases/JoinPhase.java (hunk 2 above)
Comment (matriv) [reply sql/src/main/java/io/crate/execution/dsl/phases/JoinPhase.java]: Yeah, I agree now that I see it again.
But question, do you think would be nice to have a HashJoinPhase that inherits from NestedLoopPhase? or introduce abstract JoinPhase as supe rclass for both?

File: sql/src/main/java/io/crate/execution/dsl/phases/MergePhase.java (hunk 4 above)
Comment (matriv) [reply sql/src/main/java/io/crate/execution/dsl/phases/MergePhase.java]: any suggestion where to move this method?

File: sql/src/main/java/io/crate/execution/dsl/phases/JoinPhase.java (hunk 2 above)
Comment (andreidan) [reply sql/src/main/java/io/crate/execution/dsl/phases/JoinPhase.java]: IMO there should be 2 separate components (NLPhase and HJPhase) without any inheritance

File: sql/src/main/java/io/crate/planner/operators/HashJoin.java [hunk 13]
```diff
@@ -22,42 +22,182 @@
...
+                               @Nullable Integer pageSizeHint,
+                               Row params,
+                               Map<SelectSymbol, Object> subQueryValues) {
+
+        ExecutionPlan left = lhs.build(
+            plannerContext, projectionBuilder, NO_LIMIT, 0, null, null, params, subQueryValues);
+        ExecutionPlan right = rhs.build(
+            plannerContext, projectionBuilder, NO_LIMIT, 0, null, null, params, subQueryValues);
+
+        // Move smaller table to the left side
```
Comment (andreidan) [inline sql/src/main/java/io/crate/planner/operators/HashJoin.java]: should we maybe have a method instead of this comment?

File: sql/src/main/java/io/crate/planner/operators/HashJoin.java [hunk 14]
```diff
@@ -22,42 +22,182 @@
...
+        MergePhase leftMerge = null;
+        MergePhase rightMerge = null;
+        left.setDistributionInfo(DistributionInfo.DEFAULT_BROADCAST);
+        if (JoinOperations.isNeeded(nlExecutionNodes, leftResultDesc, false)) {
+            leftMerge = JoinOperations.buildForJoin(plannerContext, leftResultDesc, nlExecutionNodes);
+        }
+        if (nlExecutionNodes.size() == 1
+            && nlExecutionNodes.equals(rightResultDesc.nodeIds())
+            && !rightResultDesc.hasRemainingLimitOrOffset()) {
+            // if the left and the right plan are executed on the same single node the mergePhase
```
Comment (andreidan) [inline sql/src/main/java/io/crate/planner/operators/HashJoin.java]: would it be more readable to have the `if` condition extracted into an `executeLeftAndRightPlanOnSameNode` method? (and move the comment inside this method)

File: sql/src/main/java/io/crate/execution/dsl/phases/JoinPhase.java (hunk 2 above)
Comment (matriv) [reply sql/src/main/java/io/crate/execution/dsl/phases/JoinPhase.java]: @andreidan I don't agree because then I need to have to Join execution plans...
https://github.com/crate/crate/pull/6903/files#diff-670704da4c49feaa815e6966562d085e

File: sql/src/main/java/io/crate/planner/operators/HashJoin.java (hunk 13 above)
Comment (matriv) [reply sql/src/main/java/io/crate/planner/operators/HashJoin.java]: the method would be weird... Would have to return a `Tuple<ExecutionPlan, ExecutionPlan>` as `left` and `right` are local variables...

File: sql/src/main/java/io/crate/planner/operators/HashJoin.java (hunk 14 above)
Comment (matriv) [reply sql/src/main/java/io/crate/planner/operators/HashJoin.java]: I don't think it adds any value. If the `if` is extracted then the comment becomes somehow irrelevant on just a boolean condition.

File: sql/src/main/java/io/crate/execution/dsl/phases/JoinPhase.java [hunk 15]
```diff
@@ -95,6 +98,8 @@ public JoinPhase(ExecutionPhase.Type type,
         this.numLeftOutputs = numLeftOutputs;
         this.numRightOutputs = numRightOutputs;
         this.leftOutputTypes = leftOutputTypes;
+        Preconditions.checkArgument(leftSize <= Integer.MAX_VALUE, "leftSize must be <= max integer value");
```
Comment (seut) [inline sql/src/main/java/io/crate/execution/dsl/phases/JoinPhase.java]: I think we should throw a more concrete exception here with a better message as this will be bubbled to the user. something like "hash join not possible as the table contains too much rows..."

File: sql/src/main/java/io/crate/planner/operators/ExplainLogicalPlan.java [hunk 16]
```diff
@@ -170,7 +170,17 @@ public Context(PlannerContext plannerContext, ProjectionBuilder projectionBuilde
...
 
         @Override
-        public ImmutableMap.Builder<String, Object> visitJoin(Join logicalPlan, Context context) {
+        public ImmutableMap.Builder<String, Object> visitNestedLoopJoin(NestedLoopJoin logicalPlan, Context context) {
+            return createMap(logicalPlan, createSubMap()
+                .put("left", explainMap(logicalPlan.lhs, context))
+                .put("right", explainMap(logicalPlan.rhs, context))
+                .put("joinType", logicalPlan.joinType())
+                .put("joinCondition",
+                    SymbolPrinter.INSTANCE.print(logicalPlan.joinCondition(), SymbolPrinter.Style.FULL_QUALIFIED)));
```
Comment (seut) [inline sql/src/main/java/io/crate/planner/operators/ExplainLogicalPlan.java]: could be extracted and used by both join plans

File: sql/src/test/java/io/crate/planner/operators/JoinTest.java (hunk 9 above)
[the remark this answers is not in the document]
Comment (seut) [reply sql/src/test/java/io/crate/planner/operators/JoinTest.java]: ok right

File: sql/src/main/java/io/crate/planner/operators/ExplainLogicalPlan.java (hunk 16 above)
Comment (matriv) [reply sql/src/main/java/io/crate/planner/operators/ExplainLogicalPlan.java]: hm no.... No superclass to call the lhs, rhs, joinType, etc...

File: sql/src/main/java/io/crate/execution/dsl/phases/HashJoinPhase.java [hunk 17]
```diff
@@ -0,0 +1,147 @@
...
+import java.util.List;
+import java.util.UUID;
+
+public class HashJoinPhase extends JoinPhase {
+
+    private final List<Symbol> leftJoinConditionInputs;
+    private final List<Symbol> rightJoinConditionInputs;
+
+    private final Collection<DataType> leftOutputTypes;
+    private final int leftSize;
```
Comment (mfussenegger) [inline sql/src/main/java/io/crate/execution/dsl/phases/HashJoinPhase.java]: maybe this could be renamed to `leftNumRowsEstimate` or something like that, to make clear it's not accurate

File: sql/src/main/java/io/crate/execution/dsl/phases/HashJoinPhase.java [hunk 18]
```diff
@@ -0,0 +1,147 @@
...
+    }
+
+    public HashJoinPhase(StreamInput in) throws IOException {
+        super(in);
+
+        leftJoinConditionInputs = Symbols.listFromStream(in);
+        rightJoinConditionInputs = Symbols.listFromStream(in);
+
+        int numLeftOutputTypes = in.readVInt();
+        if (numLeftOutputTypes > 0) {
```
Comment (mfussenegger) [inline sql/src/main/java/io/crate/execution/dsl/phases/HashJoinPhase.java]: Maybe it would make sense to add a  `DataTypes.listFromStream` as well? - Doesn't have to be in this PR, or even on this branch. Just a suggestion.

File: sql/src/main/java/io/crate/execution/dsl/phases/JoinPhase.java [hunk 19]
```diff
@@ -0,0 +1,203 @@
...
+    private final MergePhase rightMergePhase;
+    protected final JoinType joinType;
+
+    @Nullable
+    private final Symbol joinCondition;
+
+    private DistributionInfo distributionInfo = DistributionInfo.DEFAULT_BROADCAST;
+
+    JoinPhase(UUID jobId,
+                     int executionNodeId,
```
Comment (mfussenegger) [inline sql/src/main/java/io/crate/execution/dsl/phases/JoinPhase.java]: indentation is off

**Comment by matriv** [comment]
retest this please
